feat: add parsers support

cmdoret opened the pull request:
This PR adds a new "Parser" interface that can parse file contents into triples. It also migrates the existing license detection code into a parser.

This required semantic changes throughout the code, to replace "source" by "git provider" to avoid confusion.

The PR does the following:
* Move `gimie.source` -> `gimie.extractors`
* Define `Parser` interface
* Add `LicenseParser`
* Simplify logic in `Project` and make it compatible with `Parser`s
* Update all tests and docstrings
* Update CLI to:
  + Allow the user to include / exclude specific parsers in `gimie data`
  + Add a command to list available parsers.

The important changes are in `gimie/project.py`, `gimie/parsers` and `gimie/cli.py`.

Implementing a new Parser is as simple as defining a subclass in `gimie.parsers` (and adding it to [PARSERS](https://github.com/SDSC-ORD/gimie/blob/feat/parsers/gimie/parsers/__init__.py#L23)):

```py
from gimie.graph.namespaces import SDO
from gimie.parsers import Parser
from rdflib import Graph, Literal, URIRef

class ExampleParser(Parser):
    """A dummy parser for demonstration.
    It will parse files named 'example' and generate a
    graph with a single triple: 
    <URI> <schema:description <file contents>
    """

    def __init__(self, uri: URIRef):
        super().__init__(uri)

    def can_parse(resource: Resource) -> bool:
        # Match based on filename (or content)
        return Resource.name == "example"

    def _parse(resource: Resource) -> Graph:
        # Parsing logic
        contents = resource.open().read()
        return Graph().add(
            (self.uri, SDO.description, Literal(contents))
        )

```

Example usage of parsers on the command line :

```sh
# Use all parsers (default)
➜ gimie data <url>
# Only use license parser
➜ gimie data --include-parser license <url>
# Use all parsers except license
➜ gimie data --exclude-parser license <url>
# The include/exclude flags have short forms and can be specified multiple times
# Enable only cff and pyproject parsers (NOTE: these don't exist yet)
➜ gimie data -I cff -I pyproject <url> 

# list available parsers and their functionalities
➜ gimie parsers --verbose
license: Parse LICENSE file(s) into schema:license.
```

vancauwe commented on gimie/cli.py:
```diff
@@ -32,9 +34,7 @@ def cli():
-class SerializationFormat(str, Enum):
-    """Enumeration of valid RDF serialization formats for project graphs"""
-
+class FormatChoice(str, Enum):
```
could we explicit the name further for it to be standalone: LinkedDataFormatChoice or LDFormatChoice ?

cmdoret replied:
Tried to compromise between clarity and conciseness in [a6fb7e7](https://github.com/SDSC-ORD/gimie/pull/97/commits/a6fb7e71cab547e9a537d84dfc3bd0d327141a71)

vancauwe commented on gimie/project.py:
```diff
@@ -43,8 +39,12 @@ class Project:
+        ('git', 'github', 'gitlab')
+    parser_names:
+        Names of file parsers to use. ('license').
+        If None, all parsers are used.
```
This works atm for the License Parser. Based on future parsers, we might want to some parsers optional and others by default

cmdoret replied:
Agreed. but as this is not yet an issue, perhaps we can keep it for a separate PR? :grin:

cmdoret replied:
Added `DEFAULT_PARSERS` and `PARSERS`.
All functions now use `DEFAULT_PARSERS` by default.

vancauwe commented on gimie/project.py:
```diff
@@ -68,79 +71,41 @@ def __init__(
-            extractors.append(extractor)
+        files = self.extractor.list_files()
+        parsed_graphs = [
+            parser.parse_multiple(files) for parser in self.parsers
```
would it make sense separating the identification of the file from the parsing of the file. 
I.e. for files: look for license file, if there is one, call the parser. 
the question is: is it the job of a parser to find a file ? 
what if there are thousand of files being handed over from parser to parser when all the parser cares about is the small license file?

cmdoret replied:
Excellent idea! We can use a [simple factory](https://stackoverflow.com/a/45394287/8440675) here.
It could be either a `get_parser(filename)` function or a `Parser.get_parser(filename)` static method.

cmdoret replied:
`gimie.parsers.select_parser(path) -> Parser` is now available to retrieve the correct parser based on input file path.

vancauwe commented on gimie/project.py:
```diff
@@ -68,79 +71,41 @@ def __init__(
 
-    def get_extractors(self, sources: Iterable[str]) -> List[Extractor]:
+        repo = self.extractor.extract()
+        repo_graph = repo.to_graph()
```
For the repo, we extract it then make it into a graph. 
For the license, or any parsed file, shouldn't we do the same? 
Suggested process: 
1. The Parser just gets the file/files (see my comment below for file finder vs parser) and returns `SDO.license, URIRef(license_url)`. 
2. Then this gets combined with the uri into a graph (also because for other parsers it would be the same need to combine the extracted information to the uri into a graph, so we could reuse it for other parsers). 
3. Then the parser graph gets combined with the repo graph

cmdoret replied:
I see your point, we can indeed use an intermediate tuple representation like `set[tuple[URIRef, URIRef | Literal]]`. 

This would rid Parsers of the unrelated URI and make the `parse()` logic slightly simpler. Let's give it a try!

cmdoret replied:
I applied the following changes to enable this:

1. parsers now return a `Set[Tuple[URIRef, Union[URIRef, Literal]]]` as mentioned.
For readability, a type alias is defined in `gimie.graph`:
```
Property: TypeAlias = Tuple[URIRef, Union[URIRef, Literal]]
```
This allows to write signatures as `parse(bytes) -> Set[Property]`

2. A helper function `gimie.graph.operations.properties_to_graph()` helps generating a graph from these tuples.

Since Parsers only need the file contents, they have been decoupled from the `Resource` concept and only take raw data as input. This should make them easier to use in other context.

cmdoret commented:
Here is a summary of new changes:

* CLI: print parser name in bold green :sparkles:
* More flexible `PARSERS` using NamedTuple
* Helper functions to list and use parsers (thus decoupling the file matching from parsers)
* Parsers Output `Property` instead of `Graph`, helper function in gimie.graph to make the conversion
* Moved code related to parsers and extractors out of `project.py` into their respective modules
* added tests for `gimie.parsers`

vancauwe commented on gimie/graph/operations.py:
```diff
@@ -16,15 +16,24 @@
     into a single graph."""
     return reduce(lambda g1, g2: g1 | g2, graphs)
+
+
```
very nice :)

vancauwe commented on gimie/parsers/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    return get_parser(name)
+
+
+def parse_files(
```
very nice :D !

vancauwe reviewed (approved):
Nothing more to say: I really like how you chose to implement my suggestions and I find it makes the whole much more modular.
Really nice! ⭐
